trinity/tools/utils/async_generator_utils.py

```diff
@@ -0,0 +1,21 @@
+from typing import (
+    AsyncIterable,
+    List,
+)
+
+
+async def contains_all(async_gen: AsyncIterable[str], keywords: List[str]) -> bool:
+    """
+    Check wether an ``AsyncIterable[str]`` contains all of the given keywords. The keywords
+    can be embedded in some larger string. Return ``True`` as soon as all keywords were matched.
+    If not all keywords were matched by the time that the async iterable is done, return ``False``.
+    """
+    left_keywords = keywords
+    async for line in async_gen:
+        for check in left_keywords:
+            if check in line:
+                left_keywords.remove(check)
```

Comment:
I'm a fan of an alternate pattern that you may like.

```
original_values = [1, 2, 3]
seen_values = set()
for thing in things:
    if thing in set(original_values).difference(seen_things):
        seen_things.add(thing)
```

Things I like about this.

- maintains list of what we've seen (explicitely)
- doesn't require duplicating and mutating the original list.